### PR TITLE
[WIP] Adding Index Dictionary

### PR DESCRIPTION
**Pull request by kescobo**
Based on [this discussion](https://discourse.julialang.org/t/metagraph-node-indexing-w-metadata/6439) on Discourse - this PR would add a dictionary-based index to `MetaGraph` and `MetaDiGraph` types. Currently, this works:

```julia
g = MetaGraph(4)

for i in 1:4
    set_prop!(g, i, :name, "node$i")
end
update_index(g)

g["node1"] # 1
```

Open questions:

- How do deal with defaults? Using `:name` as the default index as I do now might be breaking for some people.
- How (if?) to enforce uniqueness in the indexing field? Related: how to keep the index current?

Probably lots of other things. If this seems like it would be useful, please let me know and I'll try to build it out. 

Current code is a proof of principle, but I tested it using the same setup as in the discourse topic, and got the following benchmark:

```
BenchmarkTools.Trial: 
  memory estimate:  29.04 MiB
  allocs estimate:  376480
  --------------
  minimum time:     35.263 ms (12.47% GC)
  median time:      40.081 ms (17.56% GC)
  mean time:        40.320 ms (19.28% GC)
  maximum time:     50.030 ms (29.63% GC)
  --------------
  samples:          124
  evals/sample:     1
```

Compared with the following for using filters:

```
BenchmarkTools.Trial: 
  memory estimate:  22.68 GiB
  allocs estimate:  160517950
  --------------
  minimum time:     16.994 s (32.95% GC)
  median time:      16.994 s (32.95% GC)
  mean time:        16.994 s (32.95% GC)
  maximum time:     16.994 s (32.95% GC)
  --------------
  samples:          1
  evals/sample:     1
```

[See here](https://gist.github.com/kescobo/967413b5f75ca7c6c0702b0760e80e5d) for full code and results.

### CODE REVIEW

**Comment by sbromberger** [comment]
This is a really cool way of doing it. I had been thinking about the possibility of multiple indices, though, and I don't know whether this works for that use case.

For discussion: what about a dictionary of dictionaries that enforce value uniqueness? (something akin to named sets, I guess?) Then you could have `get_vertex_by_index(g, :indexname, indexval)`, where `indexval` can be of whatever type is appropriate.

**Comment by kescobo** [comment]
Oh, that's an interesting idea. I was focused on the ability to do eg `g["node 1"]`, but in principle you could set it up so that it's sort of like indexing a 2D array. Eg you could have `g[indexval, :indexname]` be syntax sugar for `get_vertex_by_index(g, :indexname, indexval)`...

**Comment by sbromberger** [comment]
> Eg you could have `g[indexval, :indexname]` be syntax sugar for `get_vertex_by_index(g, :indexname, indexval)`

Wow. That's an interesting way of looking at it. I'm glad we're brainstorming this.

**Comment by kescobo** [comment]
```julia
using MetaGraphs

g = MetaGraph(4)

for i in 1:4
    set_prop!(g, i, :name, "node$i")
    i % 2 == 0 ? set_prop!(g, i, :even, "no!") : set_prop!(g, i, :even, "yes!")
end

set_indexing_prop!(g, :name)
g["node1", :name] # 1

set_indexing_prop!(g, :even) # Error: Cannot make even an index, duplicate values detected
g["no!", :even] # Error: ':even' is not an index
```

**Comment by kescobo** [comment]
For some reason, my recent pushes don't seem to be showing up on this PR... They can be seen [here](https://github.com/kescobo/MetaGraphs.jl/tree/dictindex)

**Comment by kescobo** [comment]
Also put in default values:

```julia
set_prop!(g, 3, :test, "mytest")
set_indexing_prop!(g, :test)
g["mytest", :test] # 3
g["test1", :test] # 1
```

**Comment by kescobo** [comment]
Some new checking:

```julia
using MetaGraphs

g = MetaGraph(4)

for i in 1:4
    set_prop!(g, i, :name, "node$i")
end

set_indexing_prop!(g, :name)
g["node1", :name] # 1

set_prop!(g, 3, :name, "name3") # Error: ':name' is an indexing property, use `set_indexing_prop!()` instead

set_indexing_prop!(g, 3, :name, "name3") # √
set_indexing_prop!(g, 3, :name, "node3") # √
set_indexing_prop!(g, 3, :name, "node3") # √
set_indexing_prop!(g, 3, :name, "node1") # Error: ':name' index already contains node1
```

**Comment by kescobo** [comment]
@sbromberger do you like where this is going? The other option is to just make a new IndexedMetaGraph type or something.

**Comment by cjprybol** [comment]
Curious spectator here 👋 . I think I've been doing the same thing as @kescobo where every time I add nodes to a LightGraph I add an entry to a dictionary mapping the name I'd like to assign to the node to the numerical index used by LightGraphs. I'm trying to learn to work with nucleotide sequence graphs, so in my case this is `Dict(DNA k-mer => node #)` and I query this external dictionary to convert between the DNA space and the LightGraphs space. I'm still learning my way through the LightGraphs and BioJulia libraries so I'm not really sure if this is the best or the recommended way to do this, but if conceptually I understand the goal of what's going on here then I'd like to give a +1 to the use-case and offer to help if there's anything useful I can contribute. Thanks to you both for your efforts here!

**Comment by sbromberger** [comment]
>  do you like where this is going?

Yes, but I confess that I haven't been able to devote much time to exploring the code yet. I am not going to be able to do a serious review of this until next week – many apologies.

**Comment by kescobo** [comment]
@sbromberger no worries! I just didn't want to go down a rabbit hole if you weren't a fan. And actually, I'm actually starting to think it might be better to make a new type, and then give the option of converting a MetaGraph into IndexableMetaGraph... I'll play with that idea this weekend or next.

**Comment by kescobo** [comment]
@cjprybol thanks for the offer! Once it's a bit more baked, some testing to see if it works for your use case would be great. and help with docs is always appreciated 😀

**Comment by sbromberger** [comment]
@kescobo - back from trip now and ready to dig in. What do benchmarks look like pre- and post- this change?

**Comment by kescobo** [comment]
Benchmarks still approx. the same as first post - the checks to see if the key exists already are quite fast. 

```julia
using LightGraphs
using MetaGraphs
using BenchmarkTools

function metadict_assign!(mgraph::MetaGraph, spindex::Vector{Int}, meindex::Vector{Int}, weights::Vector{Float64})
    for i in 1:10000
        spi = mgraph["sp$(spindex[i])", :name]
        mei = mgraph["me$(meindex[i]-500)", :name]
        weight = weights[i]
        (has_prop(mgraph, Edge(spi, mei), :weight) && weight <= get_prop(mgraph, Edge(spi, mei), :weight)) && continue

        set_prop!(mgraph, Edge(spi, mei), :weight, weight)

    end
end

function set_up()
    G = MetaGraph(1000)
    sp = Dict("sp$i" => i for i in 1:500)
    me = Dict("me$i" => i+500 for i in 1:500)

    for i in 1:1000
        if i <= 500
            set_indexing_prop!(G, i, :name, "sp$i")
        else
            set_indexing_prop!(G, i, :name, "me$(i-500)")
        end
    end

    x = rand(1:500, 10000)
    y = rand(501:1000, 10000)
    w = rand(10000)

    return (G, x, y, w)
end

function test_metadict()
    (G, x, y, w) = set_up()
    metadict_assign!(G, x, y, w)
end

@benchmark test_metadict()

#=
BenchmarkTools.Trial: 
  memory estimate:  29.52 MiB
  allocs estimate:  398029
  --------------
  minimum time:     39.560 ms (15.06% GC)
  median time:      47.135 ms (17.50% GC)
  mean time:        47.181 ms (19.92% GC)
  maximum time:     56.870 ms (17.21% GC)
  --------------
  samples:          107
  evals/sample:     1
=#
```

**Comment by sbromberger** [comment]
Nice. OK. Can you fix it so that it passes tests? (You'll need to rewrite the tests relating to constructors and you'll probably need to fix the persistence code as well so that we're saving the new information).

**Comment by kescobo** [comment]
Yes - that's the plan. Though I'm wondering if it wouldn't be better to instead have separate types for metagraphs that are indexable. I'm not going to have time to look at it again this weekend, but I'm hoping to make progress on one or the other approaches then.

**Comment by sbromberger** [comment]
> Though I'm wondering if it wouldn't be better to instead have separate types for metagraphs that are indexable

This is an option, but if there are no tradeoffs or performance regressions, there's really no reason not to include it in as a MetaGraph/MetaDiGraph.

**Comment by sbromberger** [comment]
@kescobo - you still interested in this? I am :)

S.

**Comment by kescobo** [comment]
@sbromberger Definitely - just been busy with lab deadlines etc. Because the plotting of graphs isn't quite mature, I had to shift focus to do stuff in cytoscape. 

If it's a priority, I'm 100% fine for someone else to take over. But it's on my agenda to return when I get some free time.

**Comment by sbromberger** [comment]
>  But it's on my agenda to return when I get some free time.

Awesome. I just wanted to make sure you were still interested. We'll go on your schedule.

**Comment by nickeubank** [comment]
This seems super useful - -any updates on status of this @kescobo  / need any help?

**Comment by kescobo** [comment]
@nickeubank Thanks for the bump. Alas, no updates on my end. Too many other things have had higher priority over the last 2 months. I was hoping to return to this over the holiday week, as I'll have some free time, but can't guarantee it. 

I really do plan on making progress, but I also have no problem with someone else taking over this PR if they are more likely to get it done.

**Comment by nickeubank** [comment]
@kescobo sounds good! Totally understand re: free time. :) I'm debugging / fixing a few things at the moment, but will check back in post-holidays and chip in if still needs it!

**Comment by kescobo** [comment]
@sbromberger Do you have preferences about merging vs rebasing? This PR is now behind `master` by quite a few commits.

**Comment by sbromberger** [comment]
> Do you have preferences about merging vs rebasing

I usually just do a merge, but have no real preference. Do what works best for you :)

**Comment by sbromberger** [comment]
Very happy to see this PR resurrected. Thank you!!!

**Comment by kescobo** [comment]
Alright, I think this is in pretty good shape:

```julia
julia> using MetaGraphs

julia> G = MetaGraph(100)
{100, 0} undirected Int64 metagraph with Float64 weights defined by :weight (default weight 1.0)

julia> for i in 1:100
           set_prop!(G, i, :name, "node$i")
       end

julia> set_indexing_prop!(G, :name)
Set(Symbol[:name])

julia> G["node4", :name]
4

julia> G[4, :name]
"node4"

julia> set_prop!(G, 3, :name, "name3")
ERROR: ':name' is an indexing property, use `set_indexing_prop!()` instead
Stacktrace:
 [1] set_prop!(::MetaGraphs.MetaGraph{Int64,Float64}, ::Int64, ::Symbol, ::String) at /Users/kev/.julia/v0.6/MetaGraphs/src/MetaGraphs.jl:262
```

All of the above works for `MetaDiGraphs` as well. 

@sbromberger I'm not sure if it's confusing, but I set it up so that you can do `G[i, prop]`, where `i` is an `Int`, and it will return the value of that indexing property for vertex `i`. But if the values of the indexing properties are already `Int`s, a user might expect it to return the index of the vertex where `i` is the value. In other words:

```julia
julia> for i in 1:100
           set_prop!(G, i, :times5, i*5)
       end

julia> set_indexing_prop!(G, :times5)
Set(Symbol[:name, :times5])

julia> G[5,:times5] # (1?)
25
```

I tried setting it up so that `getindex()` would check the type of the indexing property and then behave differently if it's an int, but that was far too complicated (and requires enforcing that indexing properties have a single type rather than `Any`, which opened a whole 'nother can of worms. I think as long as it's well documented it should be fine.

@cjprybol - would love it if you took a look and see if it fits your use-case. You should be able to `set_indexing_prop!(g, 3, dna"ATTAG", :kmer)`, pass the dict you already have to `set_prop!()`, and then `set_indexing_prop!(g, :kmer)` to covert. 

@nickeubank I wouldn't mind an extra set of eyes, or suggestions for how to improve the interface.

**Comment by sbromberger** [comment]
> I'm not sure if it's confusing, but I set it up so that you can do G[i, prop], where i is an Int, and it will return the value of that indexing property for vertex i.

As long as this is explained in docs, I think we're ok. It met my intuitive understanding based on the code snippet, so that's a good sign.

What happens if you try `set_indexing_prop!` on a nonunique set? (Do we have tests for that?) Also, is it possible to have multiple indexes? (I assume yes.)

**Comment by kescobo** [comment]
>What happens if you try set_indexing_prop! on a nonunique set?

Throws an error:

```julia
julia> using MetaGraphs

julia> G = MetaGraph(100)
{100, 0} undirected Int64 metagraph with Float64 weights defined by :weight (default weight 1.0)

julia> for i in 1:100
           set_prop!(G, i, :not_unique, "$(i%5)")
       end

julia> set_indexing_prop!(G, :not_unique)
ERROR: Cannot make not_unique an index, duplicate values detected
Stacktrace:
 [1] #set_indexing_prop!#2(::Void, ::Function, ::MetaGraphs.MetaGraph{Int64,Float64}, ::Symbol) at /Users/kev/.julia/v0.6/MetaGraphs/src/MetaGraphs.jl:308
 [2] set_indexing_prop!(::MetaGraphs.MetaGraph{Int64,Float64}, ::Symbol) at /Users/kev/.julia/v0.6/MetaGraphs/src/MetaGraphs.jl:306
```

Also if you try to change a value within an indexing prop to an already used value:

```
julia> set_indexing_prop!(G, 1, :name, "node1")
Set(Symbol[:name])

julia> set_indexing_prop!(G, 2, :name, "node1")
ERROR: ':name' index already contains node1
Stacktrace:
 [1] set_indexing_prop!(::MetaGraphs.MetaGraph{Int64,Float64}, ::Int64, ::Symbol, ::String) at /Users/kev/.julia/v0.6/MetaGraphs/src/MetaGraphs.jl:326
```

The latter is tested for now, I can add a test for the former... just a sec

Also, yes, you can have as many indexes as you like - it's now stored as a `Dict`

**Comment by sbromberger** [comment]
BRILLIANT.

**Comment by kescobo** [comment]
OK, tested the former case now.

File: src/MetaGraphs.jl [hunk 1]
```diff
@@ -213,7 +235,9 @@ edge `e` (optionally referenced by source vertex `s` and destination vertex `d`)
 """
 set_props!(g::AbstractMetaGraph, d::Dict) = merge!(g.gprops, d)
 set_props!(g::AbstractMetaGraph, v::Integer, d::Dict) =
-    if !_hasdict(g, v)
+    if length(intersect(keys(d), g.indices)) != 0
+        error("The following poperties are indexing_props and cannot be updated: $(intersect(keys(d), g.indices))")
```
Comment (sbromberger) [inline src/MetaGraphs.jl]: properties? (typo)

File: src/MetaGraphs.jl [hunk 2]
```diff
@@ -254,6 +283,58 @@ rem_prop!(g::AbstractMetaGraph, e::SimpleEdge, prop::Symbol) = delete!(g.eprops[
 
 rem_prop!(g::AbstractMetaGraph{T}, u::Integer, v::Integer, prop::Symbol) where T = rem_prop!(g, Edge(T(u), T(v)), prop)
 
+
+function default_index_value(v::Integer, prop::Symbol, index_values::Set{Any}; exclude=nothing)
```
Comment (sbromberger) [inline src/MetaGraphs.jl]: Can we get a docstring for this?

**Comment by sbromberger** [comment]
Two minor things and it'll be ready to merge (I can do it today). We can then work on 100% codecov (not entirely this PR's fault).

File: src/MetaGraphs.jl (hunk 2 above)
Comment (kescobo) [reply src/MetaGraphs.jl]: Sure, though this isn't an exported function ATM - do you think it should be? It's used by `set_indexing_prop!()` if not all of the values have been set yet.

**Comment by kescobo** [comment]
All set?

**Comment by sbromberger** [comment]
THIS. IS. FANTASTIC. Thank you, @kescobo, for all your work on this, and @nickeubank for the prodding :)

**Comment by kescobo** [comment]
My pleasure, glad to help! 

Thanks for sticking with me through the months of inactivity 😛

**Comment by nickeubank** [comment]
Amazing work, thanks @kescobo !

**Pull request merged**